### PR TITLE
CORE,CLI,GUI: Added new large attribute value types

### PR DESCRIPTION
**Pull request by zlamalp**
- Allow to read BLOB from SQL external source. 
  It's converted to base64 string before storing value in Perun.
- Added LargeString and LargeArrayList attribute value types.
- Handle them like String and ArrayList in Java/GUI, but
   in DB store them like large attributes (TEXT/CLOB column).
- Fixed places, where non existing java class could throw exception
  or behavior diverse by attribute type.
- Search by value is not supported for these attribute types.

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java [hunk 1]
```diff
@@ -4247,7 +4250,8 @@ public boolean isTrulyRequiredAttribute(PerunSession sess, Resource resource, Gr
 	}
 
 	public Object stringToAttributeValue(String value, String type) throws InternalErrorException {
-		if (type.equals(ArrayList.class.getName()) || type.equals(LinkedHashMap.class.getName())) {
+		if (type.equals(ArrayList.class.getName()) || type.equals(LinkedHashMap.class.getName()) ||
+				type.equals("java.util.LargeArrayList")) {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java]: I would prefer constant there.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java [hunk 2]
```diff
@@ -4318,7 +4322,7 @@ public void mergeAttributesValues(PerunSession sess, User user, List<Attribute>
 	public Attribute mergeAttributeValue(PerunSession sess, User user, Attribute attribute) throws InternalErrorException, WrongAttributeValueException,
 				 WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
 					 // Check type ArrayList
-					 if (attribute.getType().equals(ArrayList.class.getName())) {
+					 if (attribute.getType().equals(ArrayList.class.getName()) || attribute.getType().equals("java.util.LargeArrayList")) {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java]: I would prefer constant there.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java [hunk 3]
```diff
@@ -6329,7 +6333,8 @@ public Attribute convertEmptyStringIntoNullInAttrValue(Attribute attributeToConv
 		//if attributeToConvert has already null value, return it
 		if(attributeToConvert.getValue() == null) return attributeToConvert;
 		String testAttributeType = attributeToConvert.getType();
-		if (testAttributeType.equals(String.class.getName()) && attributeToConvert.getValue().equals("")) {
+		if ((testAttributeType.equals(String.class.getName()) || testAttributeType.equals("java.lang.LargeString"))
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java]: I would prefer constant there.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 4]
```diff
@@ -3331,7 +3331,9 @@ public boolean isFromNamespace(PerunSession sess, AttributeDefinition attribute,
 	}
 
 	public boolean isLargeAttribute(PerunSession sess, AttributeDefinition attribute) {
-		return attribute.getType().equals(LinkedHashMap.class.getName());
+		return (attribute.getType().equals(LinkedHashMap.class.getName()) ||
+				attribute.getType().equals("java.lang.LargeString") ||
+				attribute.getType().equals("java.util.LargeArrayList"));
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: I would prefer constant there.

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java [hunk 5]
```diff
@@ -2575,7 +2575,7 @@ private void storeApplicationAttributes(Application app) throws PerunException {
 					if (a.getType().equalsIgnoreCase("java.util.LinkedHashMap")) {
 						// FIXME do not set hash map attributes - not supported in GUI and registrar
 						continue;
-					} else if (a.getType().equalsIgnoreCase("java.util.ArrayList")) {
+					} else if (a.getType().equalsIgnoreCase("java.util.ArrayList") || a.getType().equalsIgnoreCase("java.util.LargeArrayList")) {
```
Comment (stavamichal) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java]: I would prefer constant there.

File: perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/UsersManagerMethod.java [hunk 6]
```diff
@@ -646,11 +646,11 @@ public User call(ApiCaller ac, Deserializer parms) throws PerunException {
 
 					if(attr.getType().equals(Integer.class.getName())) {
 						attr.setValue(parms.readInt("attributeValue"));
-					} else if(attr.getType().equals(String.class.getName())) {
+					} else if(attr.getType().equals(String.class.getName()) || attr.getType().equals("java.lang.LargeString")) {
```
Comment (stavamichal) [inline perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/UsersManagerMethod.java]: I would prefer constant there.

File: perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/UsersManagerMethod.java [hunk 7]
```diff
@@ -646,11 +646,11 @@ public User call(ApiCaller ac, Deserializer parms) throws PerunException {
...
 						attr.setValue(parms.readString("attributeValue"));
 					} else if(attr.getType().equals(Boolean.class.getName())) {
 						attr.setValue(parms.readBoolean("attributeValue"));
-					} else if(attr.getType().equals(ArrayList.class.getName())) {
+					} else if(attr.getType().equals(ArrayList.class.getName()) || attr.getType().equals("java.util.LargeArrayList")) {
```
Comment (stavamichal) [inline perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/UsersManagerMethod.java]: I would prefer constant there.

File: perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/client/UiElements.java [hunk 8]
```diff
@@ -403,11 +403,11 @@ static public void cantSaveAttributeValueDialogBox(Attribute object) {
 
 		String text = "";
 		String name = object.getDisplayName();
-		if (object.getType().equalsIgnoreCase("java.lang.String")) {
+		if (object.getType().equalsIgnoreCase("java.lang.String") || object.getType().equalsIgnoreCase("java.lang.LargeString")) {
```
Comment (stavamichal) [inline perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/client/UiElements.java]: I would prefer constant there.

File: perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/client/UiElements.java [hunk 9]
```diff
@@ -403,11 +403,11 @@ static public void cantSaveAttributeValueDialogBox(Attribute object) {
...
 			text = WidgetTranslation.INSTANCE.cantSaveAttributeValueDialogBoxWrongString(name);
 		} else if (object.getType().equalsIgnoreCase("java.lang.Integer")) {
 			text = WidgetTranslation.INSTANCE.cantSaveAttributeValueDialogBoxWrongInteger(name);
-		} else if (object.getType().equalsIgnoreCase("java.util.ArrayList")) {
+		} else if (object.getType().equalsIgnoreCase("java.util.ArrayList") || object.getType().equalsIgnoreCase("java.util.LargeArrayList")) {
```
Comment (stavamichal) [inline perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/client/UiElements.java]: I would prefer constant there.

File: perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/model/Attribute.java [hunk 10]
```diff
@@ -209,7 +209,7 @@ public final native boolean setValue(String newValue) /*-{
 				return false;
 			}
 		}
-		if (this.type == "java.util.ArrayList") {
+		if (this.type == "java.util.ArrayList" || this.type == "java.util.LargeArrayList") {
```
Comment (stavamichal) [inline perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/model/Attribute.java]: I would prefer constant there. (even if this is just comment)

File: perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/tabs/attributestabs/CreateAttributeDefinitionTabItem.java [hunk 11]
```diff
@@ -153,6 +153,8 @@ public boolean validateTextBox() {
 		typeListBox.addItem("Boolean", "java.lang.Boolean");
 		typeListBox.addItem("Array", "java.util.ArrayList");
 		typeListBox.addItem("LinkedHashMap", "java.util.LinkedHashMap");
+		typeListBox.addItem("LargeString", "java.lang.LargeString");
+		typeListBox.addItem("LargeArrayList", "java.util.LargeArrayList");
```
Comment (stavamichal) [inline perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/tabs/attributestabs/CreateAttributeDefinitionTabItem.java]: I would prefer constant there.

File: perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/widgets/cells/PerunAttributeValueCell.java [hunk 12]
```diff
@@ -243,7 +243,7 @@ private boolean setNewValue(Attribute attr) {
 		if (attr.getType().equals("java.lang.Boolean")) {
 			return generateValueFromBoolean(attr, getUniqueCellId(attr));
 		}
-		if (attr.getType().equals("java.util.ArrayList")) {
+		if (attr.getType().equals("java.util.ArrayList") || attr.getType().equals("java.util.LargeArrayList")) {
```
Comment (stavamichal) [inline perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/widgets/cells/PerunAttributeValueCell.java]: I would prefer constant there.

File: perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/widgets/cells/PerunAttributeValueCell.java [hunk 13]
```diff
@@ -401,7 +401,7 @@ private static String getWidget(Attribute attr) {
 			return generateNumberBox(attr.getValue(), attr.isWritable());
 		} else if (attr.getType().equals("java.lang.Boolean")) {
 			return generateCheckBox(attr.getValue(), attr.isWritable());
-		} else if (attr.getType().equals("java.util.ArrayList")) {
+		} else if (attr.getType().equals("java.util.ArrayList") || attr.getType().equals("java.util.LargeArrayList")) {
```
Comment (stavamichal) [inline perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/widgets/cells/PerunAttributeValueCell.java]: I would prefer constant there.

File: perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/model/Attribute.java (hunk 10 above)
Comment (zlamalp) [inline perun-web-gui/src/main/java/cz/metacentrum/perun/webgui/model/Attribute.java]: @stavamichal This is only usage in plain javascript, references in java are using constant.

**Pull request merged**